### PR TITLE
feat: Splitting the route by org units

### PR DESCRIPTION
**Pull request by chathurawidanage**
### Summary

The previous implementation splits the route only by the program indicators. This PR splits the route by both program indicators and the organisation units. This is done by,

1. Initially fetching OUs and assigning to a route property
2. Implementing a new MultidimensionSplitter which generates all possible pairs of OUs and PIs.
3. Analytics call is done for every possible split generated in the above step
4. Continues the old route 

Additionally,

Log4j2 was added to the project.

### CODE REVIEW

File: src/main/java/org/hisp/dhis/integration/t2a/processors/AnalyticsGridQueryBuilder.java [hunk 1]
```diff
@@ -29,25 +29,35 @@
...
+import org.apache.logging.log4j.Logger;
+import org.hisp.dhis.integration.t2a.model.OrganisationUnit;
 import org.hisp.dhis.integration.t2a.model.ProgramIndicator;
-import org.hisp.dhis.integration.t2a.routes.T2ARouter;
+import org.hisp.dhis.integration.t2a.model.Split;
 import org.hisp.dhis.integration.t2a.routes.T2ARouterNew;
 
 public class AnalyticsGridQueryBuilder implements Processor
 {
+    private static final Logger LOG = LogManager.getLogger( AnalyticsGridQueryBuilder.class );
```
Comment (mortenoh) [inline src/main/java/org/hisp/dhis/integration/t2a/processors/AnalyticsGridQueryBuilder.java:41]: Maybe you can use `@Log4j2` annotation for this? like we do in `dhis2-core` (also make sure that the log4j included is not the vulnerable one)

File: src/main/java/org/hisp/dhis/integration/t2a/routes/MultiDimensionSplitter.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+import static org.hisp.dhis.integration.t2a.routes.T2ARouterNew.PROPERTY_ALL_ORG_UNITS;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.camel.Exchange;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.hisp.dhis.integration.t2a.model.*;
```
Comment (mortenoh) [inline src/main/java/org/hisp/dhis/integration/t2a/routes/MultiDimensionSplitter.java:38]: Please don't use import collapse (we usually put it to 99 in our IDEs)

**Review by mortenoh — APPROVED** [review]

**Review by claudemamo — CHANGES_REQUESTED** [review]
This PR is duplicating my work which I had agreed to implement with @bobjolliffe.

**Review by bobjolliffe — CHANGES_REQUESTED** [review]
I think this will be too fine grained.  It is good to split along the OU dimension, but by each OU is too much.  The problem we foresee is a very large set (0.5 million sites).  Without splitting the resultset will be too large,  But splitting on each OU will be too many requests.  My view would be that Split.java needs to have not a OrganisationUnit but a Collection<OrganisationUnit>.  So we can have configurable chunks ... eg 100 or 1000 orgunits at a time.

**Comment by chathurawidanage** [comment]
I agree bob. This change can be easily incorporated. I'll wait till we merge @claudemamo 's PR. Then I'll rebase on main branch and remove redundant stuff and implement grouping if it's not available.

One question though. Should we worry about character limit for GET requests when sending multiple ous? I guess it ranges from min of 2048 to few thousands and depends on how it's configured. Usually it's safer to aim at 2048 right?

**Comment by chathurawidanage** [comment]
Closing the PR since this is obsolete.

**Pull request closed without merging**